harmonica/forward/tesseroid.py

```diff
@@ -0,0 +1,280 @@
+"""
+Forward modelling for tesseroids
+"""
+import numpy as np
+from numba import jit
+from numpy.polynomial.legendre import leggauss
+
+from ..constants import GRAVITATIONAL_CONST
+from .point_mass import jit_point_masses_gravity, kernel_potential, kernel_g_radial
+
+STACK_SIZE = 100
+GLQ_DEGREES = [2, 2, 2]
+MAX_DISCRETIZATIONS = 100000
+DISTANCE_SIZE_RATIO_POTENTIAL = 1
+DISTANCE_SIZE_RATIO_ACCELERATION = 2.5
+
+
+def tesseroid_gravity(
+    coordinates,
+    tesseroid,
+    density,
+    field,
+    glq_degrees=GLQ_DEGREES,
+    stack_size=STACK_SIZE,
+    max_discretizations=MAX_DISCRETIZATIONS,
+    three_dimensional_adaptive_discretization=False,
+):
+    """
+    Compute gravitational field of a tesseroid on a single computation point
+
+    Parameters
+    ----------
+    coordinates: list or 1d-array
+        List or array containing `longitude`, `latitude` and `radius` of a single
+        computation point defined on a spherical geocentric coordinate system.
+        Both `longitude` and `latitude` should be in degrees and `radius` in meters.
+    tesseroid : list or 1d-array
+        Geocentric spherical coordinates of the tesseroid: `w`, `e`, `s`, `n`, `bottom`,
+        `top`.
+        The longitudinal and latitudinal boundaries should be in degrees, while the
+        radial ones must be in meters.
+    density : float
+        Density of the single tesseroid in kg/m^3.
+    field: str
+        Gravitational field that wants to be computed.
+        The available fields are:
+
+        - Gravitational potential: ``potential``
+        - Radial acceleration: ``g_radial``
+    """
+    kernels = {"potential": kernel_potential, "g_radial": kernel_g_radial}
+    if field not in kernels:
+        raise ValueError("Gravity field {} not recognized".format(field))
+    # Get value of D (distance_size_ratio)
+    if field == "potential":
+        distance_size_ratio = DISTANCE_SIZE_RATIO_POTENTIAL
+    elif field == "g_radial":
+        distance_size_ratio = DISTANCE_SIZE_RATIO_ACCELERATION
+    # Convert coordinates and tesseroid to array to make Numba run only on Numpy arrays
+    tesseroid = np.array(tesseroid)
+    coordinates = np.array(coordinates)
+    # Initialize arrays to perform memory allocation only once
+    stack = np.empty((stack_size, 6))
+    small_tesseroids = np.empty((max_discretizations, 6))
+    # Apply adaptive discretization on tesseroid
+    n_splits, error = _adaptive_discretization(
+        coordinates,
+        tesseroid,
+        distance_size_ratio,
+        stack,
+        small_tesseroids,
+        radial_discretization=three_dimensional_adaptive_discretization,
+    )
+    if error == -1:
```

Comment:
@leouieda I think we should move this error rising to `_adaptive_discretization`.
Firstly because Numba supports Exceptions raising: [Numba Supported Python features](https://numba.pydata.org/numba-doc/dev/reference/pysupported.html#constructs).
Secondly, if we leave the error handling on the high-level functions, we will have to copy-paste these lines on every high-level function. That will be annoying and will have to make test functions for each time it appears.

What do you think?

Reply:
I've added a `raise ValueError` on [L289](https://github.com/santisoler/harmonica/blob/6e6e47b76ff448a97e9afc6ef7fa04d6d3bc6bfb/harmonica/forward/tesseroid.py#L289) and Numba does not complain about it.
I'm proposing something like that inside `_adaptive_discretization`.